src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/bolts/FlowSyncHubBolt.java

```diff
@@ -0,0 +1,234 @@
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.tuple.Values;
+
+public class FlowSyncHubBolt extends HubBolt implements FlowSyncCarrier, FlowPathCarrier {
+    public static final String BOLT_ID = ComponentId.FLOW_SYNC_HUB.name();
+
+    private final FlowSyncConfig config;
+    private final FlowResourcesConfig flowResourcesConfig;
+
+    private transient FlowSyncService syncService;
+    private transient FlowPathService pathService;
+
+    private final CarrierContext<String> carrierContext = new CarrierContext<>();
+    private LifecycleEvent deferredShutdownEvent;
+
+    public FlowSyncHubBolt(
+            @NonNull FlowSyncConfig config, @NonNull PersistenceManager persistenceManager,
+            @NonNull FlowResourcesConfig flowResourcesConfig) {
+        super(persistenceManager, config);
+        this.config = config;
+        this.flowResourcesConfig = flowResourcesConfig;
+    }
+
+    @Override
+    protected void onRequest(Tuple input) throws PipelineException {
+        FlowSyncRequest request = pullValue(input, FIELD_ID_PAYLOAD, FlowSyncRequest.class);
+        carrierContext.apply(pullKey(input), key -> handleRequest(key, request));
+    }
+
+    @Override
+    protected void onWorkerResponse(Tuple input) throws PipelineException {
+        SpeakerFlowSegmentResponse response = pullValue(input, FIELD_ID_PAYLOAD, SpeakerFlowSegmentResponse.class);
+        carrierContext.apply(pullKey(input), key -> handleWorkerResponse(key, response));
+    }
+
+    @Override
+    protected void onTimeout(String key, Tuple tuple) throws PipelineException {
+        syncService.handleTimeout(key);
+    }
+
+    @Override
+    protected boolean deactivate(LifecycleEvent event) {
+        if (syncService.deactivate()) {
+            return true;
+        }
+        deferredShutdownEvent = event;
+        return false;
+    }
+
+    @Override
+    protected void activate() {
+        syncService.activate();
+    }
+
+    // -- private --
+
+    private void handleRequest(String serviceKey, FlowSyncRequest request) {
+        syncService.handleRequest(serviceKey, request, getCommandContext());
+    }
+
+    private void handleWorkerResponse(String workerKey, SpeakerFlowSegmentResponse response) {
+        carrierContext.apply(KeyProvider.getParentKey(workerKey), serviceKey -> {
+            try {
+                // FlowSyncService do not communicate with speaker, so only FlowPathService can consume speaker
+                // responses.
+                pathService.handleSpeakerResponse(serviceKey, response);
+            } catch (UnknownKeyException e) {
+                log.warn("Received a response with unknown key {}.", serviceKey);
+            }
+        });
+    }
+
+    private String newPathServiceKey(PathId pathId) {
+        return KeyProvider.joinKeys(pathId.toString(), carrierContext.get());
+    }
+
+    // -- carrier --
+
+    @Override
+    public void sendSpeakerRequest(SpeakerRequest command) {
+        // TODO
```

Comment:
Why don't we send commands to FL? If we send commands, then I can't find where we do it.

Reply:
lost during seriest of rebases - fixed